src/envs/tools.py

```diff
@@ -0,0 +1,643 @@
+            "FastenScrewWithScrewdriver",
+            types=[
+                self._robot_type, self._screw_type, self._screwdriver_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenScrewByHand = ParameterizedOption(
+            # variables: [robot, screw]
+            # params: []
+            "FastenScrewByHand",
+            types=[self._robot_type, self._screw_type, self._contraption_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenNailWithHammer = ParameterizedOption(
+            # variables: [robot, nail, hammer]
+            # params: []
+            "FastenNailWithHammer",
+            types=[
+                self._robot_type, self._nail_type, self._hammer_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenBoltWithWrench = ParameterizedOption(
+            # variables: [robot, bolt, wrench]
+            # params: []
+            "FastenBoltWithWrench",
+            types=[
+                self._robot_type, self._bolt_type, self._wrench_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        # Objects
+        self._robot = Object("robby", self._robot_type)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        next_state = state.copy()
+        x, y, is_pick, is_place = action.arr
+        held = self._get_held_object(state)
+        if is_place > 0.5:
+            # Handle placing
+            if held is None:
+                # Failure: not holding anything
+                return next_state
+            on_table = self._get_contraption_pose_is_on(state, x, y) is None
+            if self._is_tool(held) and not on_table:
+                # Failure: cannot place a tool on a contraption
+                return next_state
+            next_state.set(held, "is_held", 0.0)
+            next_state.set(held, "pose_x", x)
+            next_state.set(held, "pose_y", y)
+            if self._is_item(held):
+                next_state.set(held, "on_table", float(int(on_table)))
+            next_state.set(self._robot, "fingers", 1.0)
+            return next_state
+        target = self._get_closest_object(state, x, y)
```

Comment:
can we rename this to make it more clear somehow that a contraption is not eligible to be returned?